REP: Coordinate Frames for Serial Industrial Manipulators (WIP)

gavanderhoorn opened the pull request:
Draft PR to facilitate review in ROS-I community.

Rendered version: [REP-199](http://gavanderhoorn.github.io/rep/rep-0199.html).

gavanderhoorn committed:
Initial version of 'Coordinate Frames for Serial Industrial Manipulators'.

gavanderhoorn commented:
`199` is not the official/final REP number, but a placeholder until this REP is assigned an official document ID.

jschleicher commented:
Thanks for your effort!

The orientation definition in REP 103 only says x="forward" and z="up" (very brief!). I've interpreted that for the flange differently than you do. Why should the flange outwards direction be the x-axis? To be consistent with the base frame, imho the "outwards" direction of the flange should be the z-axis.

An image in the REP would be good for clarification:
![rviz_panda-frames](https://user-images.githubusercontent.com/29709205/56958214-f5ac3880-6b49-11e9-88d4-1bd9d9f59165.png)

![rviz_prbt-frames](https://user-images.githubusercontent.com/29709205/56958314-3efc8800-6b4a-11e9-8136-d259ef2cae7c.png)

Why didn't you choose [Denavit-Hartenberg convention](https://en.wikipedia.org/wiki/Denavit%E2%80%93Hartenberg_parameters), where the joint rotation is always along the z-axis?

Example implementations:
```
# your proposal:
roslaunch abb_irb2400_support test_irb2400_12_155.launch

# flange in z-direction:
roslaunch panda_moveit_config demo.launch rviz_tutorial:=true
roslaunch prbt_support test_urdf.launch
```

jschleicher commented on rep-0199.rst:
```diff
@@ -0,0 +1,407 @@
++------------------+------------+---------------+
+|                  | This REP   | Vendor Name   |
++==================+============+===============+
+| ABB              | ``base``   | Base          |
```
I'd keep this list out of the REP and put it into the wiki. The wiki page is easier to maintain for future manufacurers and changes in vendor-specific implementations. Here a link to the wiki page would suffice.

gavanderhoorn replied:
>  The wiki page is easier to maintain for future manufacurers and changes in vendor-specific implementations

While I agree that a wiki would be easier to edit, the number of changes to this table will probably be very few. Industrial robot OEMs are not likely to change the coordinate frames they use in their controllers (because of legacy, documentation, conventions, training, etc), so this table is only expected to grow, not change.

However, I'm not entirely sold on having this table in the REP either, so I'll count your comment as a :-1: and wait for some more feedback.

jschleicher commented on rep-0199.rst:
```diff
@@ -0,0 +1,407 @@
+base_link
+'''''''''
+
+The coordinate frame called ``base_link`` shall be rigidly attached to the robot root body.
```
Which frames are optional and which ones mandatory? What happens, if a robot is missing some of the frames? Are there default assumtions such as `base=base_link`, `tool0=flange`? This would allow easier transitions of existing robot models.

```suggestion
The robot model shall define a  coordinate frame called ``base_link``.
```

gavanderhoorn replied:
> Which frames are optional and which ones mandatory?

I'll update the REP, but only `base`, `flange` and `tool0` are mandatory. The `link_n` frame names are a result of an earlier review of the REP in which we discussed whether introducing naming conventions for intermediate frames (ie: those of the links) would make sense. In the end I decided that, yes, it does make sense -- for the reasons stated in the rationale paragraph of the `link_n` section.

> What happens, if a robot is missing some of the frames?

similar to what would happen if you intend to reuse the navigation stack with a robot that does not have an `odom` frame: things will fail, as the navstack requires a REP-105 compliant system.

> Are there default assumtions such as `base=base_link`, `tool0=flange`? This would allow easier transitions of existing robot models.

While the idea of a default is nice, I'm curious how you see this working. The REP attempts to standardise urdf/xacro `link` / TF frame names. There is no facility in either of those for defaults or fallbacks.

I would also not be very comfortable if there were: without these frames present, a TF lookup `base -> tool0` could evaluate down to `base_link -> flange` (if `flange` exists), which is an entirely different lookup and the result would be completely different, semantically, from the intended lookup (as `base_link` could be anywhere and has no orientation constraints and `flange` intentionally does not correspond to `tool0`).

Finally: as I wrote above: the only frames that are *really* required are `base` and `tool0` (and potentially `flange`). If we go with the idea of defaults, just about every robot model could start claiming compliance with this REP after adding a `link` with the name `flange` somewhere. That is not the intention of the REP.

---

These are good questions btw, and I'll see how I can incorporate them (and their answers) into the REP proper.

gavanderhoorn commented:
@jschleicher thanks for your comments.

> The orientation definition in REP 103 only says x="forward" and z="up" (very brief!). I've interpreted that for the flange differently than you do. Why should the flange outwards direction be the x-axis? To be consistent with the base frame, imho the "outwards" direction of the flange should be the z-axis.

In short: because `flange` is intended to make attaching EEF sub-urdfs/xacros easier. It's not meant to be a frame that corresponds to anything that a real controller might have or define.

As ROS (ie: urdf/xacro) attaches to X+ by default (if/when you specify an identity transform), using anything but X+ makes little sense for such an attachment point (using Z+ would mean that EEFs would end up at a 90 degree angle wrt the flange by default, which would have to be undone by introducing additional counter-rotations).

Note the difference between `flange` and `tool0`: the former is purely a convenience frame, while the latter is the complement to `base` (ie: exempt from REP-103, must correspond to whatever the robot controller defines as an all-zeros toolframe).

> An image in the REP would be good for clarification:

Yes, I've already put that on my TODO list for this REP. Just haven't had time to create some nice ones.

> Why didn't you choose [Denavit-Hartenberg convention](https://en.wikipedia.org/wiki/Denavit%E2%80%93Hartenberg_parameters), where the joint rotation is always along the z-axis?

Because D-H is not very often used in ROS, not even with industrial robots. D-H is also more limited than the transforms that ROS uses and there is no need to restrict the axis of rotation to any one particular axis in URDF or XACRO.

There are actually also large industrial robot manufacturers that don't use D-H.

Can you explain why you'd like to see D-H used? Is there something that you feel would be better expressed / conveyed if we did use D-H?

Note that the REP is not about coordinate frame notation systems, but about naming and assigning semantics to frames commonly used in ROS models of industrial robots.

> Example implementations:
> 
> ```
> # your proposal:
> roslaunch abb_irb2400_support test_irb2400_12_155.launch
> ```

Please note that the ABB packages right now don't include a `flange` frame.

> ```
> # flange in z-direction:
> roslaunch panda_moveit_config demo.launch rviz_tutorial:=true
> roslaunch prbt_support test_urdf.launch
> ```

Edit: I feel like you're using `flange` here where it should actually be `tool0`. No robot's I've encountered let you plan with anything but an (unconfigured/default) tool frame (KUKA being an exception perhaps). If users don't define a tool frame, then a default should be used. But `flange` is not a tool frame. It's the name of the physical attachment point for EEFs on the robot, which are two different concepts.

jschleicher commented:
> As ROS (ie: urdf/xacro) attaches to X+ by default (if/when you specify an identity transform), using anything but X+ makes little sense for such an attachment point (using Z+ would mean that EEFs would end up at a 90 degree angle wrt the flange by default, which would have to be undone by introducing additional counter-rotations).

I didn't get that yet. If I attach some ROS grippers to flange with z+ pointing outwards, it looks like
![rviz_prbt_with_grippers](https://user-images.githubusercontent.com/29709205/57063170-a50c1b00-6cc2-11e9-805e-5bb4f820a614.png)

```
  <xacro:include filename="$(find rh_p12_rn_description)/urdf/rh_p12_rn.xacro" />

  <joint name="ee_joint" type="fixed">
    <parent link="prbt_flange"/>
    <child link="rh_p12_rn_base"/>
    <origin xyz="0. -0.1 0" /> <!-- offset for visualization to display multiple grippers -->
  </joint>

  <!-- shadow robot -->
  <xacro:include filename="$(find sr_description)/hand/xacro/three_fingers_hand.urdf.xacro" />
  <xacro:shadowhand prefix="sr_"
  muscletrans="false" muscle="false" bio="false" bt_sp="false" ubi="false" eli="false" reflect="1"/>

  <joint name="ee_joint2" type="fixed">
    <parent link="prbt_flange"/>
    <child link="sr_forearm"/>
    <origin xyz="0. 0.1 0" />
  </joint>
```

Some (most?) grippers are defined with z+ pointing inwards (as you suggest for base_link of the serial manipulator itself) and can thus be directly attached to a flange z-axis pointing outwards.

The grippers should have common naming as well, but that's probably for a next REP.

gavanderhoorn commented:
> > As ROS (ie: urdf/xacro) attaches to X+ by default (if/when you specify an identity transform), using anything but X+ makes little sense for such an attachment point (using Z+ would mean that EEFs would end up at a 90 degree angle wrt the flange by default, which would have to be undone by introducing additional counter-rotations).
> 
> I didn't get that yet. If I attach some ROS grippers to flange with z+ pointing outwards, it looks like

I haven't checked in detail as the shadowhand is quite complex, so it could be that it's rotating some of its links, or there are two effects counteracting each other in the `prbt` and the hand model.

In any case, afaict this seems to demonstrate the `x+ to x+` attachment behaviour ([gist](https://gist.github.com/gavanderhoorn/f1e1f42c5d6e68e024c3efd3157b9be5), identity transform between `tool0` (Z+ forward) and `eef` (X+ forward), with a `0.1` offset in X direction):

![rviz_screenshot_2019_05_02-12_22_32](https://user-images.githubusercontent.com/4550046/57069593-574cde00-6cd5-11e9-9498-205eccb9c493.png)

This may be due to the particular way the `tool0` frame is defined by Fanuc though (as it includes two rotations).

> Some (most?) grippers are defined with z+ pointing inwards (as you suggest for base_link of the serial manipulator itself) 

If by "inwards" you mean "somewhere between the fingers" then I would agree.

I'm not sure I follow what you mean by "as you suggest for base_link`"?

> and can thus be directly attached to a flange z-axis pointing outwards.

perhaps you've defined your `flange` frame transform differently, but that is not what I've experienced, hence the introduction of the `flange` frame with X+ always in the direction in which it would make sense to attach an EEF.

> The grippers should have common naming as well, but that's probably for a next REP.

Other REP: probably. On the other hand: if the gripper follows REP-103 *and* the model starts with a `base_link` things *should* work out without having to standardise anything else.

gavanderhoorn commented:
@jschleicher: OK, I just checked the `prbt` setup.

The fact that we don't see the issue there is because there is no X or Z offset between the `prbt`'s `flange` and the EEF. Whatever the EEF does internally doesn't matter, as it's essentially a black box.

The idea of `flange` as proposed by this REP is to be able to always treat the robot model as a black box as well and allow users to assume that "forward" is always in the X+ direction. Attaching an EEF where it's (for some reason or another) necessary to include some offsets "in the forward" direction would make having a frame where "forward" is always in the X+ direction actually desirable.

Irrespective of what the zero pose of the robot is (ie: fully stretched out along X+, fully stretched out along Z+ or some combination), `flange` would always have X+ "forward", meaning that any distance to be covered between the physical flange and the physical EEF would correspond to an offset in the X+ direction.

So without knowing anything about the robot model and how it's setup, I can always assume that attaching an end-effector is done along a REP-103 aligned frame.

An alternative would be to say that Z+ could be used as well (which may be what you implied with the question "why not use D-H?"). And that is true. It would mean deviating from the "X+ is forward" convention of REP-103. If the idea is similar, but there is one alternative that aligns with a previous standard/REP/convention, I believe we should use that alternative.

---

Edit: I believe I now also understand your comment:

> The orientation definition in REP 103 only says x="forward" and z="up" (very brief!). I've interpreted that for the flange differently than you do. Why should the flange outwards direction be the x-axis? 

REP-103 says this:

> ## Axis Orientation
> 
> In relation to a body the standard is:
> 
>   - x forward

Note: "in relation to a body". So this is about link-local frames for a URDF.

To me this sounds like for all links, locally, X+ should be forward. If we then consider that most flanges (ie: physically) will be disks or disk-like shapes, the "forward" direction would make the most sense (to me, I admit) to be pointing out of the disk, perpendicular to the largest surface of the disk.

jschleicher commented:
> An alternative would be to say that Z+ could be used as well (which may be what you implied with the question "why not use D-H?"). And that is true. It would mean deviating from the "X+ is forward" convention of REP-103.

That's exactly, what I propose. 
And an offset is not necessary: For perfecet robot and gripper models you can directly link `gripper_base_link` to `robot_flange`.

Maybe REP-103 was defined with mobile robots in mind, where "forwards" and "up" are unambigious? A manipulators `up` direction depends on the position of the joints, so the only possibility is to use the "link frame"-forward and that would result in changing base_link as well:

> if the gripper follows REP-103 and the model starts with a base_link things should work out without having to standardise anything else.

:+1: 

But: For the proposed `base_link` the z-axis is perpendicular to the mounting flange. For your proposed `flange` link the x-axis would be perpendicular to the flange?!

gavanderhoorn commented:
> > An alternative would be to say that Z+ could be used as well (which may be what you implied with the question "why not use D-H?"). And that is true. It would mean deviating from the "X+ is forward" convention of REP-103.
> 
> That's exactly, what I propose.
> And an offset is not necessary: For perfecet robot and gripper models you can directly link gripper_base_link to robot_flange.

Right. But "perfect" does not exist. We cannot assume that all grippers and all robots are always perfect.

In addition, the X offset was just an example.

The main point is that by allowing arbitrary axes to be used for this we lose the abstraction that a standardised frame allows for.

I also added this after the sentence that you quote:

>  If the idea is similar, but there is one alternative that aligns with a previous standard/REP/convention, I believe we should use that alternative.

given that REP-103 prescribes such a convention, it would seem to make sense to use the alternative that aligns with that convention.

> Maybe REP-103 was defined with mobile robots in mind, where "forwards" and "up" are unambigious? A manipulators `up` direction depends on the position of the joints, [..]

I believe the important distinction here is "in relation to a body" in REP-103. We're talking about link-local frames here, so all directions are body-relative. Up is always "up", the pose of the body in some other external coordinate frame does not matter.

> so the only possibility is to use the "link frame"-forward and that would result in changing base_link as well:

---

(edit: hm, wait. Perhaps I only understand your comment now: yes, if with "link frame-forward" you mean the cardinalities as defined in the link-local system then I believe we understand each other, but I don't see how that would necessarily imply changes for `base_link`. The base of a robot is typically mounted flatly on a surface, if we then consider it's local coordinate system (ie: stand behind it) the natural or logical position as asked for by the REP would seem to work. Are you suggesting that X+ should also point "up", out of the "disk" that is typically the top of a robot's base? That would not be logical to me, but I can see why you'd say that given that the REP does propose to setup `flange` exactly like that.)

---

I don't think we can draw this conclusion, as you seem to (implicitly) assume that there is some global external reference frame that we can use to unambiguously determine what is up. There isn't and we can't.

If we'd mount a robot inverted on a ceiling things would change.

> > if the gripper follows REP-103 and the model starts with a base_link things should work out without having to standardise anything else.
> 
> +1
> 
> But: For the proposed `base_link` the z-axis is perpendicular to the mounting flange. For your proposed `flange` link the x-axis would be perpendicular to the flange?!

yes, again: because these are link-local frames.

I admit things like "forward", "left" and "right" will come down to personal preference and convention somewhat (and that is partly what the REP is trying to codify here), but if we look at a typical flange, defining "forward" as sticking out of one of its sides seems at least curious to me.

---

Btw: thanks for discussing this. These are important points and I feel talking about them helps improve the REP. Regardless of what happens to `flange`, it would appear additional clarification and/or examples are needed to get the main point(s) across, which is already very welcome input.

ted-miller commented:
> Definition for TCP, "Tool Center Point"

Generally it is best to not refer to it as "center" point. Instead it is a "control" point. (Tool Control Point) This helps to avoid confusion on large EFF devices where the TCP is often offset from the center.

> flange, "Positive X (x+) must always point away from the last link (ie: in the 'forward' direction for a world-aligned robot model)."

This sentence seems ambiguous. I think it should be expressed relative to a particular axis of the previous axis. Or, the "forward" direction should be defined relative to 'base_link'.
(The orientation for the final flange axis, relative to the world, will vary on different models of yaskawa robots.)

tool0:
I don't understand the purpose of having the 'tool0' frame. If it can't move and there's no children, then why does it exist? (Keep in mind, I don’t actually have any experience using ROS... so I may be ignorant of something fundamental.)

In your "Multi-group (asymmetric)" section, you make reference to a "world" frame. However, there is no definition for this frame in your document.

How do you handle a scenario where the robot is mounted to a linear rail? Does the origin of the rail become 'base'? I don't understand the following paragraph when it is applied to the context of a robot mounted to a rail:
> "Any frame is acceptable as the parent of base, as long as the transform between parent and base is fixed (i.e.: not across a movable joint) and the location and orientation of the frame always correspond to the controller's internal default Cartesian frame. As many serial manipulators share a common design, it is expected however that base_link will often be the parent of base, connected by a fixed joint with a suitable transform."

Nomenclature mapping:
I think you should also include "world" frame here. 
The Yaskawa vendor names should be, world = Base (BF), base = Robot (RF), tool0 = Tool (TF#). The abbreviations are what is used when programming a job on the pendant.

gavanderhoorn commented:
> > Definition for TCP, "Tool Center Point"
> 
> Generally it is best to not refer to it as "center" point. Instead it is a "control" point. (Tool Control Point) This helps to avoid confusion on large EFF devices where the TCP is often offset from the center.

I'll consider this. Is this Yaskawa-specific terminology? A quick google only turns up results from CnC literature, and the Yaskawa Robotics Glossary ([here](https://www.motoman.com/en-us/about/company/robotics-glossary)) and even that says: "See Tool Center Point".

> > flange, "Positive X (x+) must always point away from the last link (ie: in the 'forward' direction for a world-aligned robot model)."
> 
> This sentence seems ambiguous. I think it should be expressed relative to a particular axis of the previous axis. 

@jschleicher had similar comments. I'll include a figure/diagram/screenshot of the intended pose.

Please note though: this REP builds on REP-105, which standardises "up", "left" and "forward" and the positive directions of all axes in a body-relative frame. In this case the body would be the last link (ie: the flange).

> Or, the "forward" direction should be defined relative to 'base_link'. (The orientation for the final flange axis, relative to the world, will vary on different models of yaskawa robots.)

We can't use any world relative references here, as that would still be ambiguous (and depend on the zero pose of the robot).

> tool0:
> I don't understand the purpose of having the 'tool0' frame. If it can't move and there's no children, then why does it exist? (Keep in mind, I don’t actually have any experience using ROS... so I may be ignorant of something fundamental.)

Good question.

There are a number of uses, one of which would be facilitating reuse of off-line generated (ie: code generated) solvers and planners that are specific to chain configurations. IKFast solvers would be an example.

> In your "Multi-group (asymmetric)" section, you make reference to a "world" frame. However, there is no definition for this frame in your document.

You're correct. I just wanted to include an example that would be a little more like a real-world setup. I should not introduce a "world" frame there though. I'll remove it.

> How do you handle a scenario where the robot is mounted to a linear rail? Does the origin of the rail become 'base'? 

According to the REP right now, `base` should not change and must always correspond to the origin of the robot-local base coordinate system. I realise (now) there are robots where `base` (or its equivalent) *can* change. I'm not entirely sure yet, but I don't believe I would be in favour of allowing changes to base -- even in that case:

 - (a bit weak, but) this REP is about frames for industrial manipulators. It'd be difficult to consider all possible kinematics configurations, so I wanted to start with just manipulators in isolation.
 - Of course robots may be integrated with additional kinematics, but even then those robots can typically be used in a stand-alone fashion (fi: without the rail) and I'd like to support that by allowing robot-local coordinate transforms (ie: to the `base` of the robot)

The [asymmetric example](http://gavanderhoorn.github.io/rep/rep-0199.html#multi-group-asymmetric) shows a possible way to handle this: `robot_base` is the world coordinate system of the robot proper, `rail_base` would be the origin of the rails coordinate system, and a transform `rail_base->robot_tool0` should result in what I believe you are asking for/about.

That would not seem to need any changes to the REP (apart from perhaps phrasing and explaining this better).

> I don't understand the following paragraph when it is applied to the context of a robot mounted to a rail:
> 
> > "Any frame is acceptable as the parent of base, as long as the transform between parent and base is fixed (i.e.: not across a movable joint) and the location and orientation of the frame always correspond to the controller's internal default Cartesian frame. As many serial manipulators share a common design, it is expected however that base_link will often be the parent of base, connected by a fixed joint with a suitable transform."

Does my previous comment already help?

With a robot on a rail, we'd get (with a `robot_` prefix): `robot_base_link->robot_base`, and (with `rail_` prefix for the rail links): `rail_base_link->rail_base`.

> Nomenclature mapping:
> I think you should also include "world" frame here.
> The Yaskawa vendor names should be, world = Base (BF), base = Robot (RF), tool0 = Tool (TF#). The abbreviations are what is used when programming a job on the pendant.

Thanks, I'll update the names in the table.

As I wrote earlier though, I don't think we'll include the `world` frame (it's not specific to manipulators, which is the topic of this REP).

gavanderhoorn committed:
Clarify the REP follows REP-103 completely.

ahoarau commented:
Hello,
As I was explaining to one of our interns the naming for industrial manipulators frame names, I got myself a bit confused and had to go back again to this REP to re-discover the problem once again :

- Having "vendor-independent" frames for base and flange
- Having "vendored" frames for base and flange
- Having something called "tool" when it is not representing a physical object

The "base" of the robot, as we all call it, is called `base_link` here, and the vendor name is called `base`. Typical usecase when computingkinematic frames : "This pose is computed with respect to the base of the robot... called base_link. Do not use 'base' as the frame of reference"

The "Tool" pose of the robot is confusing in itself as it can represent:
- The transform between `flange-->eef_tcp` 
- The pose of the visual mesh representing the "thing attached to the robot flange"
- The pose of the tcp frame w.r.t the robot base

As `tool0` is only here as a way to transform poses into the vendored convention, why not call them `base_vendor` and `flange_vendor` ?
Also, why not call (`base_link` and `flange`)
- `base` and `flange`
or
- `base_link` and `flange_link` 
?

```shell
base
├ base_vendor  # vendored version of the base
└ link_1
  └ ..
    └ link_N
      └ flange
        └ flange_vendor # vendored version of the flange = 'Tool' pose when T(flange->eef_tcp) = Identity
```

Thanks for your comments ! 

ref: http://gavanderhoorn.github.io/rep/rep-0199.html

EDIT: This is a comment from https://github.com/ros-industrial/rep/issues/29, wrongly posted

gavanderhoorn commented:
@ahoarau: thanks for your comment.

Would I understand you correctly if I summarise it as follows:

 1. could we not use the name `base` for what is now called `base_link`?, and
 1. could we not use the name `base_vendor` for what is now called `base`?, and
 1. could we not use the name `flange_vendor` for what is now `tool0`?

If you can confirm, I will try to answer.

ahoarau commented:
Yes. That is correct.

gavanderhoorn commented:
https://github.com/gavanderhoorn/rep/blame/coord_frames_ind_robots/rep-0199.rst#L152

gavanderhoorn commented:
> Yes. That is correct.

Ok. See below.

>  1\. could we not use the name `base` for what is now called `base_link`?

The short answer would be: no, we cannot do that. The rationale being: the name `base_link` has special meaning "in ROS". `base_link` is understood/implied/taken to be "the start" of any robot model. And with start, I mean: the root of the model/xacro/urdf.

(note: this "root" is *not* necessarily the root of the kinematic model, but merely the for humans/visual/collision model)

There are quite some tools and nodes which expect (wrongly perhaps, but still) that `base_link` is the root of the robot. Both tools used and developed be the ROS community, as well as tools outside it (which happen to have added compatibility with URDF).

As such, it would be unwise the use a different name for the root of the model, and it will simplify adoption of this REP (and the standardisation which it attempts to achieve) if xacros/urdfs of robot models which comply with it conform to existing (best) practices and allow reuse of existing tooling as much as possible.

So we opted not to rename `base_link` for the reasons explained above. We could of course just use a different name (instead of "renaming `base_link`) but that would essentially cause the same issues.

This sentence from the REP tries to say the same thing, but in a very condensed form (from the end of the `base_link` section):

> As this frame is by convention considered the start of a ROS kinematic (sub) chain, using it in the same way in models of serial manipulators is strongly recommended, as this facilitates integration with other ROS tools and infrastructure.

I'll try to make this more clear, as you are not the first commenter to ask this question.

Now for the root of the kinematic tree, see my reply to your next question.

> 2\. could we not use the name `base_vendor` for what is now called `base`?

First: the main point of even having a `base` frame is to support encoding the start of the kinematic tree or model exactly as the OEM has defined it.

For many industrial robots this is *not* identical to the "bottom" or "mounting point" of the robot's mechanical structure, but has its origin somewhere else (good examples here are Fanuc and Motoman, which put the origin of their World frames at the intersection of two joint axes, whereas ABB fi does the opposite and places the origin of the World frame at the bottom (ie: mounting point) of the robot).

The transform `base_link -> base` allows us to express this difference.

Now as to your question: of course we could have used a name other than `base` for this. But that would perhaps have led to questions such as "*Could we not use the name `base` for what is now called `base_vendor`?*" In essence, this is just a name, and in this case, this name was chosen. It's concise, not too long and after reading the REP, its semantics should be clear.

And I believe that this:

> Typical usecase when computingkinematic frames : "This pose is computed with respect to the base of the robot... called base_link. Do not use 'base' as the frame of reference"

would actually be *exactly* what `base` is for: to express a pose of the TCP (or any frame really) in the Cartesian World frame of a particular robot (with ROS support), one should transform it to its `base` frame. As that frame corresponds to the World frame's location as defined by the OEM, the transform `base -> my_eef` should correspond exactly to what the real robot reports (provided the intermediate transforms are also correct of course).

I don't completely understand what you mean by this part: "*Do not use 'base' as the frame of reference*".

> 3\. could we not use the name `flange_vendor` for what is now `tool0`?

First: the idea behind `tool0` is for it to represent (or encode) the location and orientation of an *unconfigured* or *default* tool frame. Again: *exactly as the OEM has defined it* for a particular industrial robot model. For many industrial robots, this will mean it will be coincident with `flange`, or with the last physical link, if the robot does not have a `flange` frame defined (which would be in violation of this REP).

As to the name I would repeat my response to your earlier question about `base` and `base_vendor` (ie: of course a different name could be used). But in this case there is precedence for using the name "tool zero": on ABB controllers fi, `tool0` is the default tool frame. On KUKAs, `tool_data[0] == flange` IIRC. On Fanuc controllers, the "zero index" for user frame data indicates "no user frame configured". While that is not a toolframe, the idea of 0 being a default or unconfigured frame is there and we wanted to use that here to convey the same thing.

Other OEMs have similar semantics attached to these unconfigured frames and indexes.

> The "Tool" pose of the robot is confusing in itself as it can represent:
> 
> * The transform between `flange-->eef_tcp`
> 
> * The pose of the visual mesh representing the "thing attached to the robot flange"
> 
> * The pose of the tcp frame w.r.t the robot base

I hope I've been able to clarify that `tool0` has strictly defined semantics, and cannot be anything else but what the first bullet describes here.

(as part of the transform `base->tool0`, it would of course also be what your third bullet describes, but only if the first bullet is actually true *and* the default tool frame is configured as being the *active* tool frame on the robot controller).

---

The main motivation for the REP is really to introduce:

> Consistent naming of frames and standardisation of semantics

as this:

> allows for re-use of [..] software [and] also of experience and tooling

and:

> to make it possible for ROS applications to refer to [..] coordinate frames in a controller and vendor-agnostic way

By introducing these abstract names, it becomes possible to treat robot models (ie: urdf/xacros) as black boxes, as the start and end points are clearly defined, and attaching EEFs and other frames to them should no longer require to look at the actual xacro content itself. As long as these agreed upon frames are present, users may assume knowing about `base_link`, `base`, `flange` and `tool0` should be sufficient.

Summarising:

 * `base_link` is the start of the ROS model (ie: mechanical mounting point, or connection to the rest of the URDF)
 * `base` encodes the start of the kinematic chain (ie: mathematical) *and* connects the OEM specific kinematic origin to `base_link`
  * `flange` is a vendor agnostic attachment point, which always follows REP-103 (abstracts away pose differences between robots in their "zero pose" so we can avoid unwanted rotations of EEF models).
    It is also the end of the "mechanical chain" if you will (which starts with `base_link`).
  * `tool0` represents the *default* end of the kinematic chain (which starts with `base`), and represents a default or unconfigured tool frame.

Besides providing an always present default toolframe, `tool0` has additional value: there are kinematic solvers and motion planners which make use of code generation. That code generation is typically done for a specific kinematic chain which must not (actually: cannot) be changed afterwards without updating the generated code. By providing the chain `base -> tool0`, such a solver can be generated for that particular chain and can then be (re)used as part of a larger chain or tree (provided there are transforms available to and from `base` and `tool0`).

gavanderhoorn commented:
And they are also linked in the [References](http://gavanderhoorn.github.io/rep/rep-0199.html#references) section of the candidate REP, but some links which may give more insight into the discussion around `base` and `tool0`:

 - https://github.com/ros-industrial/ros_industrial_issues/issues/24
 - https://github.com/ros-industrial/abb/issues/32
 - https://github.com/ros-industrial/universal_robot/pull/200#issuecomment-102980913

ahoarau commented:
Thanks for the clear explanation. The need to have a generic base, OEM base, generic tool and OEM tool frames makes perfect sense.

> Besides providing an always present default toolframe, tool0 has additional value: there are kinematic solvers and motion planners which make use of code generation. That code generation is typically done for a specific kinematic chain which must not (actually: cannot) be changed afterwards without updating the generated code. By providing the chain base -> tool0, such a solver can be generated for that particular chain and can then be (re)used as part of a larger chain or tree (provided there are transforms available to and from base and tool0).

Why wouldn't you have your kinematic chains from `base_link` to `flange` (all the same for all robots, easy to debug/test/show in 3d), and just provide an additional transformation for the OEM frame as a different function or option to your solvers ? (like `forwardKinematics(UseOEMConventions)`)

> The short answer would be: no, we cannot do that. The rationale being: the name base_link has special meaning "in ROS". base_link is understood/implied/taken to be "the start" of any robot model. And with start, I mean: the root of the model/xacro/urdf.

I see that `base_link` has some historical/common explanation/meaning, let's keep that then.
But then let's add consistency to the rest: `flange` --> `flange_link` ?
And most of all, **semantics**: `base` --> `oem_base` and   `tool0` --> `oem_tool0` ? Basically any combination with `oem_` or `vendor` would be enough.

![image](https://user-images.githubusercontent.com/703240/79915065-96b09500-8426-11ea-830f-2a12cc4d0314.png)

Really the fact that  I'm not the only one asking questions about those is a clear indication that the semantics are not clear enough :)

gavanderhoorn commented:
> Thanks for the clear explanation. The need to have a generic base, OEM base, generic tool and OEM tool frames makes perfect sense.

Good. Seems we agree on that then.

> > Besides providing an always present default toolframe, tool0 has additional value: there are kinematic solvers and motion planners which make use of code generation. That code generation is typically done for a specific kinematic chain which must not (actually: cannot) be changed afterwards without updating the generated code. By providing the chain base -> tool0, such a solver can be generated for that particular chain and can then be (re)used as part of a larger chain or tree (provided there are transforms available to and from base and tool0).
> 
> Why wouldn't you have your kinematic chains from `base_link` to `flange` (all the same for all robots, easy to debug/test/show in 3d), 

Because that wouldn't correspond to how many industrial robot OEMs define their chains. The goals of the REP are both to help ROS users deal with idiosyncrasies from the industrial space, as well as help industrial robot OEMs map their world to ROS.

Keeping the mapping as close to what is already there (ie: decades of industrial deployments) is assumed (and has turned out) to help with that.

> and just provide an additional transformation for the OEM frame as a different function or option to your solvers ? (like `forwardKinematics(UseOEMConventions)`)

While that would certainly be possible, it would mean we'd need additional software and update existing software. Everything is possible of course, but the REP right now tries to find a balance between documenting what is there, moving towards new standards/best practices *and* allowing as much reuse of existing infrastructure as possible.

If I understand your suggestion correctly, the `base_link -> base` fixed transform seems to be doing the same, but without requiring new software or updates. TF can already do all the lookups for us, and we just add "an additional transform for the OEM frame", as you suggest. It's just that we keep in the model, instead of 'hiding' it in the code.

But there is one unresolved issue with the text in this REP and that may relate to your question: most IK solvers *are* actually generated for `base_link -> tool0`, not `base -> tool0`.

> > The short answer would be: no, we cannot do that. The rationale being: the name base_link has special meaning "in ROS". base_link is understood/implied/taken to be "the start" of any robot model. And with start, I mean: the root of the model/xacro/urdf.
> 
> I see that `base_link` has some historical/common explanation/meaning, let's keep that then.
> But then let's add consistency to the rest: `flange` --> `flange_link` ?
> And most of all, **semantics**: `base` --> `oem_base` and `tool0` --> `oem_tool0` ? Basically any combination with `oem_` or `vendor` would be enough.

To me this sounds like just another name, just as `base` and `tool0` are "just names". 

I"m not sure whether you've seen them, but the comments in https://github.com/ros-industrial/ros_industrial_issues/issues/24 go into why no `oem_` or `mfg_` or similar prefixes have been added (my own comments start from https://github.com/ros-industrial/ros_industrial_issues/issues/24#issuecomment-37762869).

Just to be clear: I'm not *entirely* against adding such a prefix (notice how in https://github.com/ros-industrial/ros_industrial_issues/issues/24 I actually suggested something similar), but I did (and do) agree with @shaun-edwards that it would be better to not have them.

Although `oem_` or `mfg_` may be clearer in the case of industrial robots where it's clear who the OEM is, that is not necessarily the case for robot (models) which would still benefit from using these standardised frames (because it makes sense for them) but which would not be industrial robots, not have an OEM or for some other reason it would not make sense to talk about an OEM or manufacturer.

If we include those prefixes in the REP, the same question you pose here (but turned around) could be asked: "*Why add `oem_` or `mfg_` or `vendor_` prefixes to these frames? I / my robot does not have an OEM or vendor; what do these prefixes refer to?*".

By using a lowest-common denominator approach (ie: it's not ubiqitous so don't include it), we end up with frame names without prefixes which make sense in all contexts (note: not perfect sense perhaps, as they could be clearer).

> ![image](https://user-images.githubusercontent.com/703240/79915065-96b09500-8426-11ea-830f-2a12cc4d0314.png)
> 
> Really the fact that I'm not the only one asking questions about those is a clear indication that the semantics are not clear enough :)

I feel you've taken that quote out of context just a little bit, but to me, the fact that more commenters have asked about it (and I did really only mean that particular aspect you commented on) seems to imply the descriptions in the REP should be improved.

Not necessarily that names should be changed.

It would be something to consider though.

ahoarau commented:
> Although oem_ or mfg_ may be clearer in the case of industrial robots where it's clear who the OEM is, that is not necessarily the case for robot (models) which would still benefit from using these standardised frames (because it makes sense for them) but which would not be industrial robots, not have an OEM or for some other reason it would not make sense to talk about an OEM or manufacturer.

If you don't have an industrial robot, then you might not use any "ros-**industrial**" stuff, so I'm guessing this is out of scope. 

> To me this sounds like just another name, just as base and tool0 are "just names".

Absolutely. I still **vote** for adding a meaningful prefix to those frames. It makes things clearer.

> Just to be clear: I'm not entirely against adding such a prefix (notice how in ros-industrial/ros_industrial_issues#24 I actually suggested something similar), but I did (and do) agree with @shaun-edwards that it would be better to not have them.

Are you refering to this comment ?
```
We could add the prefix mfg to tool0 and base but would create a lot of work to port these changes to existing URDFs and keep things backwards compatible. My preference is to stick tool0 to avoid this work. This would also mean sticking with the somewhat ambiguous base name.
```

gavanderhoorn commented:
> If you don't have an industrial robot, then you might not use any "ros-**industrial**" stuff, so I'm guessing this is out of scope.

The REP will be submitted to `ros-infrastructure/rep`, not `ros-industrial/rep`.

Similar to [REP-105](https://www.ros.org/reps/rep-0105.html) (Coordinate Frames for Mobile Platforms) and [REP-120](https://www.ros.org/reps/rep-0120.html) (Coordinate Frames for Humanoid Robots), it is intended to be usable (and used) by the wider ROS community.

There is no assumption this should apply only to industrial robots with large installed bases or robots distributed by large companies.

> > To me this sounds like just another name, just as base and tool0 are "just names".
> 
> Absolutely. I still **vote** for adding a meaningful prefix to those frames. It makes things clearer.

Noted.

> Are you refering to this comment ?
>> We could add the prefix mfg to tool0 and base but would create a lot of work to port these changes to existing URDFs and keep things backwards compatible. My preference is to stick tool0 to avoid this work. This would also mean sticking with the somewhat ambiguous base name.

Among others, yes.

The bw-compatibility concern is still relevant.

ahoarau commented:
I probably got confused with :
![image](https://user-images.githubusercontent.com/703240/80212079-2331a200-8637-11ea-921e-9ff6ac145320.png)

I totally get it if this Draft is made for "everybody using ROS" rather than "everybody using serial industrial manipulators"

gavanderhoorn commented:
> I probably got confused with :
> ![image](https://user-images.githubusercontent.com/703240/80212079-2331a200-8637-11ea-921e-9ff6ac145320.png)

Yeah, I figured that was what you were going to reply with :)

Touché.

We'll have to fix that.

> I totally get it if this Draft is made for "everybody using ROS" rather than "everybody using serial industrial manipulators"

To clarify this a bit, an example I have in my immediate work:

 - gantry system with 3 + N DoF
 - multiple 'heads'
 - changeable EEFs
 - design is custom, realisation a cooperation of multiple companies

It comes with its own controller with coordinate system conventions, world, object, tool and other such frames. The `base_link` of this thing is nowhere near the origin of its world frame (`base_link -> base` is offset by several metres). It also has multiple such transforms, as it's not a single "robot" in a sense. There are multiple `base -> tool0` chains and we have multiple `flange`s. EEFs are attached to those `flange`s. Per-project TCPs are defined as siblings of `tool0`, with `flange` as their parent.

I would clearly identify this as an industrial machine, very much like a robot (clearly not an *industrial serial manipulator* though, although they can be attached to it), for which the use of the structure put forth in this REP really helps. 

As it does not have a single OEM or mfg, and in fact, *we* decide the configuration of the coordinate systems (and the control systems in general), the `oem_`/`mfg_`/`vendor_` prefix doesn't really work. 

The *industrial serial* in the title was put there when this REP *was* targeted only at those types of robots. As use the concepts described in the REP have started to be used outside of that narrow scope, the title should be fixed.

---

Thanks for your comments btw; they've been valuable. I'm going to update this text when I get a chance, and I'll try to properly use your input.

gavanderhoorn committed:
Clarify status of the REP.

gavanderhoorn committed:
Remove vendor mapping

Difficult to verify and incorrect in some cases.

It only distracted from the content and will be hard to keep up-to-date.

It also confused people, as it makes this REP seemingly only applicable to industrial robots, which is not the intention.

gavanderhoorn committed:
Link IDs come from set of positive ints
